### PR TITLE
feat(ams): add startup optimizer bootstrap with auto-recreate optimizer

### PR DESCRIPTION
**Pull request by arsalan-datazip**
## Why are the changes needed?
Move Spark optimizer setup from Helm lifecycle hooks and Docker Compose init scripts into Fusion itself. On startup, Fusion creates and manages the optimizer resource group and Spark job from optimizer.* config. 

## How was this patch tested?
- [x] Run test locally before making a pull request

### CODE REVIEW

**Comment by badalprasadsingh** [comment]
Can you please think of these edge cases, and tell me if the solution will work out or not:

- `Spark driver pod evicted/OOM-killed` -> a replacement will be scheduled, the new pod will have the same config, and when it `authenticates`, the old token will still be in DB. Thus, two optimizers should pop up on the UI.

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    }
+  }
+
+  private void createOptimizerResource(
+      AbstractOptimizerContainer container, ResourceGroup group, int parallelism) {
+    Resource resource =
+        new Resource.Builder(group.getContainer(), group.getName(), ResourceType.OPTIMIZER)
+            .setProperties(group.getProperties())
+            .setThreadCount(parallelism)
+            .build();
```
Comment (badalprasadsingh) [inline amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:219]: Missed `.setMemorySize(...)`.

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+          "Cannot release optimizer on non-optimizer resource container %s.",
+          resource.getContainerName());
+      ((AbstractOptimizerContainer) resourceContainer).releaseResource(resource);
+      released = true;
+    } catch (Exception e) {
+      LOG.warn("Failed to release optimizer resource {}", resourceId, e);
+    }
+    if (released) {
+      cleanupResource(resource.getGroupName(), resourceId);
+    }
```
Comment (badalprasadsingh) [inline amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:260]: Here, the `finally` block always runs. Are we sure it is required to behave like this? What if the kill fails?

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+  private void createOptimizerResource(
+      AbstractOptimizerContainer container, ResourceGroup group, int parallelism) {
+    Resource resource =
+        new Resource.Builder(group.getContainer(), group.getName(), ResourceType.OPTIMIZER)
+            .setProperties(group.getProperties())
+            .setThreadCount(parallelism)
+            .build();
+    container.requestResource(resource);
+    optimizerManager.createResource(resource);
```
Comment (badalprasadsingh) [inline amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:221]: Any reason to request the resource before creating the resource?

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    } catch (Exception e) {
+      LOG.warn("Failed to delete resource {} from database", resourceId, e);
+    }
+    try {
+      optimizerManager.deleteOptimizer(groupName, resourceId);
+    } catch (Exception e) {
+      LOG.warn("Failed to delete optimizer {} from database during cleanup", resourceId, e);
+    }
+  }
+}
```
Comment (badalprasadsingh) [inline amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:282]: Isn't this redundant?

File: amoro-ams/src/main/java/org/apache/amoro/server/AmoroServiceContainer.java [hunk 5]
```diff
@@ -257,7 +260,7 @@ public void startOptimizingService() throws Exception {
     tableService.initialize();
     LOG.info("AMS table service have been initialized");
     tableManager.setTableService(tableService);
-
+    new OptimizerBootstrap(serviceConfig, optimizerManager, optimizingService).run();
```
Comment (badalprasadsingh) [inline amoro-ams/src/main/java/org/apache/amoro/server/AmoroServiceContainer.java]: Please add this into a try-catch block.

**Review by badalprasadsingh — CHANGES_REQUESTED** [review]
First Review.

Please add the modification header where-ever required.

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java (hunk 2 above)
Comment (arsalan-datazip) [reply amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:260]: fixed

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java (hunk 3 above)
Comment (arsalan-datazip) [reply amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:221]: requestResource first: doScaleOut enriches resource.getProperties() with metadata
needed by releaseResource. 
createResource then persists the complete record
before the optimizer's authenticate() API call arrives.

This is also followed while creating the resource thorugh API.

File: amoro-ams/src/main/java/org/apache/amoro/server/AmoroServiceContainer.java (hunk 5 above)
Comment (arsalan-datazip) [reply amoro-ams/src/main/java/org/apache/amoro/server/AmoroServiceContainer.java]: done

File: amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java (hunk 4 above)
Comment (arsalan-datazip) [reply amoro-ams/src/main/java/org/apache/amoro/server/bootstrap/OptimizerBootstrap.java:282]: done changes

**Comment by badalprasadsingh** [comment]
Closing this **PR** as the _direction_ of the feature has moved to _auto-creation_ of optimizer as well

EDIT: Let's not create a new PR

**Comment by badalprasadsingh** [comment]
Closing this PR as the feature direction has moved a lot.
